File: app/UI/Hammer.hs
{-
    Módulo que forja telas que contém placeholder.
-}
module Hammer where

-- | Função que faz o replace dos elementos de place holders.
-- Lê todos os caracteres da tela atual e substitui o caracter caso ele seja um place holder.
-- Espera um background, contendo place holders do caracter #, e uma String com as informações
-- a serem usadas para replace.
forgeScreen :: String ->  String -> String
forgeScreen "" _ = ""
forgeScreen (h:t) especialChar
    | h == '#' = take 1 especialChar ++ forgeScreen t (drop 1 especialChar)
    | otherwise = h : forgeScreen t especialChar

-- | Esta função faz o merge das representações contidas em uma matriz.
--  A saída está normalizada para a função `forgeScreen`.
--  Espera um inteiro que é o index da última linha das representações,
--  além de uma matriz com as representações, com o mesmo length.
mergeControll :: Int -> [[String]] -> String
mergeControll (-1) _ = ""
mergeControll lineIndex arr =  mergeControll (lineIndex - 1) arr ++ mergeLine lineIndex (length arr - 1) arr

-- | Esta função faz o merge das linhas de cada elemento da matriza, faz isso
-- utilizando como base um inteiro que representa a linha a ser olhada e que é
-- passado como argumento.
mergeLine :: Int -> Int -> [[String]] -> String
mergeLine _ (-1) _ = ""
mergeLine lineIndex idx arr = mergeLine lineIndex (idx - 1) arr ++ (arr !! idx) !! lineIndex

File: app/UI/Hammer.pl
/*
    Predicado de acesso ao processo de Replace.
 Recebe uma Lista que é a representação da tela, uma Lista que contém os elementos
 de replace e um argumento que representa a saída.
 Realiza a normalização das listas, isso significa que o predicado prepara os dois
 elementos de entrada para serem usados no predicado `forgeScreen`, não sendo 
 necessário utilizar os predicados `unlines` e `string_chars` para usar o forge.
*/
anvil(ScreenList, ReplaceList, Result) :-
    normalizedIn(ReplaceList, "", ReplaceNormalized),
    normalizedIn(ScreenList, "\n", ScreenNormalized),

    forgeScreen(ScreenNormalized, ReplaceNormalized, Out),
    Result = Out.

/*
    Predicado de substituição dos caracteres de replace.
 Realiza o replace, quando há placeholders, de informações pertinentes à tela passada.
 Recebe uma lista que contém os caracteres da tela a ser representada, uma Lista que 
 contém os elementos de replace e uma lista que será a tela forjada.
*/
forgeScreen([], L, L) :- !.
forgeScreen(BackGroundL, [], R) :-
    R = BackGroundL, !.

forgeScreen([Hback | Tback], [Hrepl | Trepl], [Hresp | Tresp]) :-
    atom_string(Hback, "#"), 
    Hresp = Hrepl,
    forgeScreen(Tback, Trepl, Tresp), !.

forgeScreen([Hback | Tback], [Hrepl | Trepl], [Hresp | Tresp]) :- 
    Hresp = Hback,
    forgeScreen(Tback, [Hrepl | Trepl], Tresp), !.


/*
    Predicado que Mescla Listas contidas em uma Lista.
 O retorno do predicado é uma lista que contém, para cada elemento da lista resposta,
 a concatenação de todos os elementos de uma determinada linha `NLine`. Esses elementos
 vêm de todas as listas internas sendo concatenadas no formato: [elem1++elem2++...++elemN].
*/
mergeControll(_, -1, []) :- !.
mergeControll(Matrix, NLine, [Hresp | Tresp]) :-
    currentNumberLine(Matrix, NLine, NumOut),
    mergeLine(Matrix, NumOut, Result),
    unlines(Result, "", NormalizedResult),

    Hresp = NormalizedResult,
    NextLine is NLine -1,
    mergeControll(Matrix, NextLine, Tresp).

% Passa por todas as listas internas e faz a concatenação dos elementos da linha `NLine`.
mergeLine([], _, []) :- !.
mergeLine([Hmatrix | Tmatrix], NLine, [Hresp | Tresp]) :-
    nth0(NLine, Hmatrix, Element),
    Hresp = Element,
    mergeLine(Tmatrix, NLine, Tresp).

% Predicado auxiliar para a correção do número da linha atual.
currentNumberLine([H | _], Line, Result) :-
    length(H, Len),
    Result is (Len - 1) - Line.

% Predicado auxiliar para a unificação dos elementos de uma lista.
unlines(Strings, Symbol, Result) :- atomic_list_concat(Strings, Symbol, Result).

/* 
 Predicado auxiliar que normaliza uma entrada contendo uma lista de 
 representações para uma lista de caracteres.
*/
normalizedIn(List, Symbol, Result) :-
    unlines(List, Symbol, ListUnlines),
    string_chars(ListUnlines, ListUnlinesChars),
    Result = ListUnlinesChars.